Infer correct trait bounds for use of a type parameter associated type

The following expands to `impl<T: Trait> Serialize for Struct<T>` which fails to compile.

```rust
trait Trait {
    type Assoc;
}

#[derive(Serialize)]
struct Struct<T: Trait> {
    assoc: T::Assoc,
}
```

The builtin derives already get this partially right. For example `#[derive(Debug)]` on the above expands to `impl<T: Debug + Trait> Debug for Struct<T> where T::Assoc: Debug`. This is partially right because it has `T::Assoc: Debug`, but wrong because it also has `T: Debug`.

I've taken a stab at implementing this [here](https://github.com/Osspial/serde/tree/assoc_type_derive), which works by explicitly adding a `Serialize`/`Deserialize` requirement to all of the fields in a derived type that go through serialization/deserialization, respectively. It seems to work on the associated types I tried, but it's preventing some of the tests from compiling, and I'm not entirely sure how to go about fixing them. Would you mind looking at my code?

```rust
error[E0275]: overflow evaluating the requirement `test_gen::RecursiveA: serde::Serialize`
   --> test_suite\tests\test_gen.rs:158:14
    |
158 |     #[derive(Serialize, Deserialize)]
    |              ^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0275`.
```

--- a/serde_derive/src/bound.rs
+++ b/serde_derive/src/bound.rs
@@ -55,21 +55,22 @@ pub fn with_where_predicates_from_fields<F, W>(
     generics: &syn::Generics,
     trait_bound: &syn::Path,
     from_field: F,
-    where_ty: W,
+    gen_bound_where: W,
 ) -> syn::Generics
 where
     F: Fn(&attr::Field) -> Option<&[syn::WherePredicate]>,
-    W: for<'a> Fn(&attr::Field) -> Option<&syn::ExprPath>,
+    W: for<'a> Fn(&attr::Field) -> bool,
 {
     let predicates = cont.data
         .all_fields()
         .flat_map(|field| {
-            let field_bound: Option<syn::WherePredicate> = match where_ty(&field.attrs) {
-                Some(_) => None,
-                None => {
+            let field_bound: Option<syn::WherePredicate> = match gen_bound_where(&field.attrs) {
+                true => {
                     let field_ty = field.ty;
-                    Some(parse_quote!(#field_ty: #trait_bound))
-                }
+                    let predicate: syn::WherePredicate = parse_quote!(#field_ty: #trait_bound);
+                    Some(predicate)
+                },
+                false => None
             };
             field_bound.into_iter().chain(from_field(&field.attrs).into_iter().flat_map(|predicates| predicates.to_vec()))
         });

--- a/serde_derive/src/bound.rs
+++ b/serde_derive/src/bound.rs
@@ -50,18 +50,29 @@ pub fn with_where_predicates(
     generics
 }
 
-pub fn with_where_predicates_from_fields<F>(
+pub fn with_where_predicates_from_fields<F, W>(
     cont: &Container,
     generics: &syn::Generics,
+    trait_bound: &syn::Path,
     from_field: F,
+    where_ty: W,
 ) -> syn::Generics
 where
     F: Fn(&attr::Field) -> Option<&[syn::WherePredicate]>,
+    W: for<'a> Fn(&attr::Field) -> Option<&syn::ExprPath>,
 {
     let predicates = cont.data
         .all_fields()
-        .flat_map(|field| from_field(&field.attrs))
-        .flat_map(|predicates| predicates.to_vec());
+        .flat_map(|field| {
+            let field_bound: Option<syn::WherePredicate> = match where_ty(&field.attrs) {
+                Some(_) => None,
+                None => {
+                    let field_ty = field.ty;
+                    Some(parse_quote!(#field_ty: #trait_bound))
+                }
+            };
+            field_bound.into_iter().chain(from_field(&field.attrs).into_iter().flat_map(|predicates| predicates.to_vec()))
+        });
 
     let mut generics = generics.clone();
     generics.make_where_clause()

--- a/serde_derive/src/bound.rs
+++ b/serde_derive/src/bound.rs
@@ -59,24 +59,27 @@ pub fn with_where_predicates_from_fields<F, W>(
 ) -> syn::Generics
 where
     F: Fn(&attr::Field) -> Option<&[syn::WherePredicate]>,
-    W: for<'a> Fn(&attr::Field) -> bool,
+    W: Fn(&attr::Field) -> bool,
 {
     let predicates = cont.data
         .all_fields()
         .flat_map(|field| {
             let field_ty = field.ty;
-            let field_bound: Option<syn::WherePredicate> = match (field_ty, gen_bound_where(&field.attrs)) {
-                (&syn::Type::Path(ref ty_path), true) => match (ty_path.path.segments.first(), generics.params.first()) {
-                    (Some(syn::punctuated::Pair::Punctuated(ref t, _)), Some(syn::punctuated::Pair::End(&syn::GenericParam::Type(ref generic_ty))))
-                        if generic_ty.ident == t.ident =>
-                    {
-                        let predicate: syn::WherePredicate = parse_quote!(#field_ty: #trait_bound);
-                        Some(predicate)
-                    },
-                    _ => None
-                },
-                (_, _) => None
+            let matching_generic = |t: &syn::PathSegment, g| match g {
+                &syn::GenericParam::Type(ref generic_ty)
+                    if generic_ty.ident == t.ident => true,
+                _ => false
             };
+
+            let mut field_bound: Option<syn::WherePredicate> = None;
+            if let &syn::Type::Path(ref ty_path) = field_ty {
+                field_bound = match (gen_bound_where(&field.attrs), ty_path.path.segments.first()) {
+                    (true, Some(syn::punctuated::Pair::Punctuated(ref t, _)))
+                        if generics.params.iter().fold(false, |b, g| b || matching_generic(t, g))
+                        => Some(parse_quote!(#field_ty: #trait_bound)),
+                    (_, _) => None
+                };
+            }
             field_bound.into_iter().chain(from_field(&field.attrs).into_iter().flat_map(|predicates| predicates.to_vec()))
         });
 

--- a/serde_derive/src/bound.rs
+++ b/serde_derive/src/bound.rs
@@ -65,18 +65,19 @@ where
         .all_fields()
         .flat_map(|field| {
             let field_ty = field.ty;
-            let matching_generic = |t: &syn::PathSegment, g| match g {
-                &syn::GenericParam::Type(ref generic_ty)
+            let matching_generic = |t: &syn::PathSegment, g: &syn::GenericParam| match *g {
+                syn::GenericParam::Type(ref generic_ty)
                     if generic_ty.ident == t.ident => true,
                 _ => false
             };
 
             let mut field_bound: Option<syn::WherePredicate> = None;
-            if let &syn::Type::Path(ref ty_path) = field_ty {
+            if let syn::Type::Path(ref ty_path) = *field_ty {
                 field_bound = match (gen_bound_where(&field.attrs), ty_path.path.segments.first()) {
-                    (true, Some(syn::punctuated::Pair::Punctuated(ref t, _)))
-                        if generics.params.iter().fold(false, |b, g| b || matching_generic(t, g))
-                        => Some(parse_quote!(#field_ty: #trait_bound)),
+                    (true, Some(syn::punctuated::Pair::Punctuated(ref t, _))) =>
+                        if generics.params.iter().any(|g| matching_generic(t, g)) {
+                            Some(parse_quote!(#field_ty: #trait_bound))
+                        } else {None},
                     (_, _) => None
                 };
             }

--- a/serde_derive/src/bound.rs
+++ b/serde_derive/src/bound.rs
@@ -64,13 +64,18 @@ where
     let predicates = cont.data
         .all_fields()
         .flat_map(|field| {
-            let field_bound: Option<syn::WherePredicate> = match gen_bound_where(&field.attrs) {
-                true => {
-                    let field_ty = field.ty;
-                    let predicate: syn::WherePredicate = parse_quote!(#field_ty: #trait_bound);
-                    Some(predicate)
+            let field_ty = field.ty;
+            let field_bound: Option<syn::WherePredicate> = match (field_ty, gen_bound_where(&field.attrs)) {
+                (&syn::Type::Path(ref ty_path), true) => match (ty_path.path.segments.first(), generics.params.first()) {
+                    (Some(syn::punctuated::Pair::Punctuated(ref t, _)), Some(syn::punctuated::Pair::End(&syn::GenericParam::Type(ref generic_ty))))
+                        if generic_ty.ident == t.ident =>
+                    {
+                        let predicate: syn::WherePredicate = parse_quote!(#field_ty: #trait_bound);
+                        Some(predicate)
+                    },
+                    _ => None
                 },
-                false => None
+                (_, _) => None
             };
             field_bound.into_iter().chain(from_field(&field.attrs).into_iter().flat_map(|predicates| predicates.to_vec()))
         });

--- a/serde_derive/src/de.rs
+++ b/serde_derive/src/de.rs
@@ -124,7 +124,9 @@ impl Parameters {
 fn build_generics(cont: &Container, borrowed: &BorrowedLifetimes) -> syn::Generics {
     let generics = bound::without_defaults(cont.generics);
 
-    let generics = bound::with_where_predicates_from_fields(cont, &generics, attr::Field::de_bound);
+    let delife = borrowed.de_lifetime();
+    let de_bound = parse_quote!(_serde::Deserialize<#delife>);
+    let generics = bound::with_where_predicates_from_fields(cont, &generics, &de_bound, attr::Field::de_bound, attr::Field::deserialize_with);
 
     match cont.attrs.de_bound() {
         Some(predicates) => bound::with_where_predicates(&generics, predicates),
@@ -136,12 +138,11 @@ fn build_generics(cont: &Container, borrowed: &BorrowedLifetimes) -> syn::Generi
                 attr::Default::None | attr::Default::Path(_) => generics,
             };
 
-            let delife = borrowed.de_lifetime();
             let generics = bound::with_bound(
                 cont,
                 &generics,
                 needs_deserialize_bound,
-                &parse_quote!(_serde::Deserialize<#delife>),
+                &de_bound,
             );
 
             bound::with_bound(

--- a/serde_derive/src/de.rs
+++ b/serde_derive/src/de.rs
@@ -126,7 +126,13 @@ fn build_generics(cont: &Container, borrowed: &BorrowedLifetimes) -> syn::Generi
 
     let delife = borrowed.de_lifetime();
     let de_bound = parse_quote!(_serde::Deserialize<#delife>);
-    let generics = bound::with_where_predicates_from_fields(cont, &generics, &de_bound, attr::Field::de_bound, attr::Field::deserialize_with);
+    let generics = bound::with_where_predicates_from_fields(
+        cont,
+        &generics,
+        &de_bound,
+        attr::Field::de_bound,
+        |field| field.deserialize_with().is_none() && !field.skip_deserializing()
+    );
 
     match cont.attrs.de_bound() {
         Some(predicates) => bound::with_where_predicates(&generics, predicates),

--- a/serde_derive/src/ser.rs
+++ b/serde_derive/src/ser.rs
@@ -130,8 +130,9 @@ impl Parameters {
 fn build_generics(cont: &Container) -> syn::Generics {
     let generics = bound::without_defaults(cont.generics);
 
+    let trait_bound = parse_quote!(_serde::Serialize);
     let generics =
-        bound::with_where_predicates_from_fields(cont, &generics, attr::Field::ser_bound);
+        bound::with_where_predicates_from_fields(cont, &generics, &trait_bound, attr::Field::ser_bound, attr::Field::serialize_with);
 
     match cont.attrs.ser_bound() {
         Some(predicates) => bound::with_where_predicates(&generics, predicates),
@@ -139,7 +140,7 @@ fn build_generics(cont: &Container) -> syn::Generics {
             cont,
             &generics,
             needs_serialize_bound,
-            &parse_quote!(_serde::Serialize),
+            &trait_bound
         ),
     }
 }

--- a/serde_derive/src/ser.rs
+++ b/serde_derive/src/ser.rs
@@ -131,8 +131,13 @@ fn build_generics(cont: &Container) -> syn::Generics {
     let generics = bound::without_defaults(cont.generics);
 
     let trait_bound = parse_quote!(_serde::Serialize);
-    let generics =
-        bound::with_where_predicates_from_fields(cont, &generics, &trait_bound, attr::Field::ser_bound, attr::Field::serialize_with);
+    let generics = bound::with_where_predicates_from_fields(
+        cont,
+        &generics,
+        &trait_bound,
+        attr::Field::ser_bound,
+        |field| field.serialize_with().is_none() && !field.skip_serializing()
+    );
 
     match cont.attrs.ser_bound() {
         Some(predicates) => bound::with_where_predicates(&generics, predicates),

--- a/test_suite/tests/test_gen.rs
+++ b/test_suite/tests/test_gen.rs
@@ -555,6 +555,14 @@ fn test_gen() {
     }
 
     assert::<AssocDerive<NoSerdeImpl>>();
+
+    #[derive(Serialize, Deserialize)]
+    struct AssocDeriveMulti<S, T: AssocSerde> {
+        s: S,
+        assoc: T::Assoc,
+    }
+
+    assert::<AssocDeriveMulti<i32, NoSerdeImpl>>();
 }
 
 //////////////////////////////////////////////////////////////////////////

--- a/test_suite/tests/test_gen.rs
+++ b/test_suite/tests/test_gen.rs
@@ -539,6 +539,22 @@ fn test_gen() {
         array: [u8; 256],
     }
     assert_ser::<BigArray>();
+
+    trait AssocSerde {
+        type Assoc;
+    }
+
+    struct NoSerdeImpl;
+    impl AssocSerde for NoSerdeImpl {
+        type Assoc = u32;
+    }
+
+    #[derive(Serialize, Deserialize)]
+    struct AssocDerive<T: AssocSerde> {
+        assoc: T::Assoc
+    }
+
+    assert::<AssocDerive<NoSerdeImpl>>();
 }
 
 //////////////////////////////////////////////////////////////////////////

